### PR TITLE
Fix #2

### LINKED ISSUE #2: Implement UART
See [d16 reference manual][], §4.1 "UART" for the implementation specification.

[d16 reference manual]: https://github.com/d16-processor/d16/blob/master/doc/D16%20Reference%20Manual.pdf

### PR DESCRIPTION
**Pull request by flaviut**

### CODE REVIEW

File: d16i/uart.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+            try:
+                c = getch()
+                self.rx_fifo.put(ord(c[0]), block=False)
+            except Queue.QueueFull:
+                self.rx_lock.acquire()
+                self.rx_overrun = True
+                self.rx_lock.release()
+
+
+def getch():
```
Comment (flaviut) [inline d16i/uart.py]: Cite your sources.

Also, how about http://stackoverflow.com/a/21659588, so it's cross-platform?

File: d16i/uart.py.bak [hunk 2]
```diff
@@ -0,0 +1,114 @@
+from sys import stdin, stdout
```
Comment (flaviut) [inline d16i/uart.py.bak]: Delete and add to gitignore.

**Comment by flaviut** [comment]
Also delete and add to gitignore `*.o`.
Squash everything down. My commit message is pretty shit, and the "fix tox" stuff doesn't really belong as a separate commit because it basically just reverts parts of other commits.

File: d16i/cpu.py [hunk 3]
```diff
@@ -256,6 +261,7 @@ def _decode_jmpsel(self):
     # }}}
 
     def execute_instruction(self):
+        # self.stats.instructions_dispatched += 1
```
Comment (flaviut) [inline d16i/cpu.py]: Delete this line entirely if it isn't be used. Also, `CpuStatistics` can be deleted if it isn't being used.

File: d16i/uart.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+    def end(self):
+        while not self.tx_fifo.empty():
+            pass
+
+    def write_uart_data(self, data: int):
+        self.tx_fifo.put(data)
+
+    def read_uart_status(self) -> int:
+        self.rx_lock.acquire()
```
Comment (flaviut) [inline d16i/uart.py]: This should use the lock as a context manager instead. Like

```python
with self.rx_lock:
    ...
```

**Comment by C-Elegans** [comment]
I really hope we don't break anything by re-merging with master....

**Comment by C-Elegans** [comment]
Ok, merged without breaking stuff. Whew.

**Pull request merged**